### PR TITLE
Migrate the yaml used by the eager flow to flow.flex.yaml, code support both flow.flex.yaml and flow.dag.yaml.

### PR DESCRIPTION
**Pull request by Stephen1993**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

1. Modify the **yaml** used for all **eager tests** to **flow.flex.yaml.**
2. Modify the **yaml** of the **eager flow** in the code to **flow.flex.yaml**.
3. Add **flow** when saving run to/retrieving run from the **database**

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/_utils/flow_utils.py [hunk 1]
```diff
@@ -75,15 +75,25 @@ def resolve_flow_path(
     else:
         flow_path = Path(flow_path)
 
+    if flow_path.is_dir():
+        target_folder = flow_path
+        target_filer = FLOW_FLEX_YAML if (target_folder / FLOW_FLEX_YAML).is_file() else FLOW_DAG_YAML
+        if (target_folder / FLOW_DAG_YAML).is_file() and (target_folder / FLOW_FLEX_YAML).is_file():
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/_utils/flow_utils.py:81]: let's refactor to only do two is_file call here

File: src/promptflow-core/promptflow/_utils/flow_utils.py [hunk 2]
```diff
@@ -75,15 +75,25 @@ def resolve_flow_path(
...
+            raise ValidationException(
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/_utils/flow_utils.py:82]: add some test to guard the behavior here

File: src/promptflow-core/promptflow/_utils/flow_utils.py (hunk 1 above)
Comment (Stephen1993) [reply src/promptflow-core/promptflow/_utils/flow_utils.py]: done

File: src/promptflow-core/promptflow/_utils/flow_utils.py (hunk 2 above)
Comment (Stephen1993) [reply src/promptflow-core/promptflow/_utils/flow_utils.py:85]: done

File: src/promptflow-devkit/promptflow/_sdk/entities/_run.py [hunk 3]
```diff
@@ -716,6 +716,7 @@ def _load_from_source(cls, source: Union[str, Path], params_override: Optional[D
             run_info = json.load(f)
 
         return cls(
+            flow=run_info.get("properties", {}).get("azureml.promptflow.flow_definition_blob_path", None),
```
Comment (wangchao1230) [inline src/promptflow-devkit/promptflow/_sdk/entities/_run.py:719]: add a constant for the property key?

Comment (wangchao1230) [inline src/promptflow-devkit/promptflow/_sdk/entities/_run.py:719]: The key should not start with azureml for local run

File: src/promptflow-core/promptflow/_utils/flow_utils.py [hunk 4]
```diff
@@ -141,20 +148,25 @@ def dump_flow_dag(flow_dag: dict, flow_path: Path):
...
-            file_path = working_dir / file_path
-        if file_path.suffix.lower() not in [".yaml", ".yml"]:
-            return False
-        yaml_dict = load_yaml(file_path)
+    if flow_path is not None:
+        flow_path, flow_file = resolve_flow_path(flow_path, base_path=working_dir, check_flow_exist=False)
+        file_path = flow_path / flow_file
+        if file_path.is_file() and file_path.suffix.lower() in (".yaml", ".yml"):
+            yaml_dict = load_yaml(file_path)
+        if not check_flow_exist and flow_file == FLOW_FLEX_YAML:
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/_utils/flow_utils.py:167]: shall we just rely on the entry check?

File: src/promptflow-devkit/promptflow/_sdk/_utils.py [hunk 5]
```diff
@@ -1028,6 +1030,22 @@ def callable_to_entry_string(callable_obj: Callable) -> str:
...
+        try:
+            # The flow yaml may have been temporarily generated and deleted after creating a run.
+            # So check_flow_exist=False
+            return is_flex_flow(flow_path=run.flow, check_flow_exist=False)
+        except Exception as e:
+            # For run with incomplete flow snapshot, ignore load flow error to make sure it can still show.
+            logger.debug(f"Failed to check is flex flow from {run.flow} due to {e}.")
+            return False
+    elif run._run_source in [RunInfoSources.INDEX_SERVICE, RunInfoSources.RUN_HISTORY]:
+        return run._properties.get("azureml.promptflow.run_mode") == "Eager"
```
Comment (wangchao1230) [inline src/promptflow-devkit/promptflow/_sdk/_utils.py:1044]: make "Eager“  an constant? change it to Flex?

File: src/promptflow-core/promptflow/_utils/flow_utils.py (hunk 4 above)
Comment (Stephen1993) [reply src/promptflow-core/promptflow/_utils/flow_utils.py:167]: This is to fix a bug
When using **pf.run**, if there is no flow file, yaml file will be temporarily generated and deleted after creating run. 
At this time, **run.flow** is a non-existent file that cannot be determined by **load_yaml** and needs to be determined by the **run.flow suffix.**

image

File: src/promptflow-devkit/promptflow/_sdk/_utils.py (hunk 5 above)
Comment (Stephen1993) [reply src/promptflow-devkit/promptflow/_sdk/_utils.py]: Done

File: src/promptflow-devkit/promptflow/_sdk/entities/_run.py (hunk 3 above)
Comment (Stephen1993) [reply src/promptflow-devkit/promptflow/_sdk/entities/_run.py]: This modification has been deleted.

Comment (Stephen1993) [reply src/promptflow-devkit/promptflow/_sdk/entities/_run.py]: This modification has been deleted.

**Review by 0mza987 — APPROVED** [review]

File: src/promptflow-devkit/promptflow/_sdk/operations/_run_operations.py [hunk 6]
```diff
@@ -360,7 +361,7 @@ def _visualize(self, runs: List[Run], html_path: Optional[str] = None) -> None:
                 metrics=local_storage.load_metrics(parse_const_as_str=True),
                 dag=local_storage.load_dag_as_string(),
                 flow_tools_json=local_storage.load_flow_tools_json(),
-                mode="eager" if local_storage.eager_mode else "",
+                mode=EAGER.lower() if local_storage.eager_mode else "",
```
Comment (elliotzh) [inline src/promptflow-devkit/promptflow/_sdk/operations/_run_operations.py:364]: this constant name is a little confusing. What about put it under a class like:
```python
class RunMode:
  EAGER = "eager"
```

File: src/promptflow-core/promptflow/_utils/flow_utils.py [hunk 7]
```diff
@@ -141,20 +148,25 @@ def dump_flow_dag(flow_dag: dict, flow_path: Path):
...
-            file_path = working_dir / file_path
-        if file_path.suffix.lower() not in [".yaml", ".yml"]:
-            return False
-        yaml_dict = load_yaml(file_path)
+    if flow_path is not None:
+        flow_path, flow_file = resolve_flow_path(flow_path, base_path=working_dir, check_flow_exist=False)
+        file_path = flow_path / flow_file
+        if file_path.is_file() and file_path.suffix.lower() in (".yaml", ".yml"):
+            yaml_dict = load_yaml(file_path)
+        elif not check_flow_exist:
```
Comment (elliotzh) [inline src/promptflow-core/promptflow/_utils/flow_utils.py:167]: why we need this parameter?

Comment (Stephen1993) [reply src/promptflow-core/promptflow/_utils/flow_utils.py:167]: This is to fix a bug
When using pf.run, if there is no flow file, yaml file will be temporarily generated and deleted after creating run.
At this time, run.flow is a non-existent file that cannot be determined by load_yaml and needs to be determined by the run.flow suffix.

image
image

File: src/promptflow-devkit/promptflow/_sdk/operations/_run_operations.py (hunk 6 above)
Comment (Stephen1993) [reply src/promptflow-devkit/promptflow/_sdk/operations/_run_operations.py:364]: Yes, I will modify it.

**Review by elliotzh — DISMISSED** [review]

**Review by 0mza987 — APPROVED** [review]

**Review by 0mza987 — APPROVED** [review]

**Pull request merged**